File: pku/2466.java
import java.util.*;
import static java.lang.Math.*;

public class Main{
    public static void main(String[] args){
        new Main().run();
    }

    void run(){
        Scanner sc = new Scanner(System.in);
        loop:
        for(;sc.hasNext();){
            int n = sc.nextInt();
            int total = n * (n-1) / 2;
            int[] ps = new int[total];
            for(int i = 0; i < total; i++)
                ps[i] = sc.nextInt();
            Arrays.sort(ps);
            boolean[] u = new boolean[total];
            u[0] = u[1] = true;
            for(int i = 2; i < total; i++){
                int a2 = ps[0] + ps[1] - ps[i];
                if(a2 % 2 == 1)continue;
                u[i] = true;
                int a = a2 / 2;
                ArrayList<Integer> cand = new ArrayList<Integer>();
                cand.add(a); cand.add(ps[0] - a); cand.add(ps[1] - a);
                boolean flag = true;
                Hashtable<Integer, Integer> ht = new Hashtable<Integer, Integer>();
                for(int j = 2; j < total; j++)
                    if(i!=j){
                        if(ht.containsKey(ps[j]))
                            ht.put(ps[j], ht.get(ps[j])+1);
                        else
                            ht.put(ps[j], 1);
                    }
                jail:
                for(int j = 3; j < n; j++){
                    int mn = 1<<30;
                    for(int k = 2; k < total; k++)
                        if(ht.containsKey(ps[k]) && ht.get(ps[k]) > 0){
                            mn = ps[k];
                            ht.put(ps[k], ht.get(ps[k])-1);
                            break;
                        }
                    int v = mn - a;
                    for(int k = 1; k < cand.size(); k++){
                        int sum = v + cand.get(k);
                        if(!ht.containsKey(sum) || ht.get(sum) <= 0){
                            flag = false;
                            break jail;
                        }
                        ht.put(sum, ht.get(sum)-1);
                    }
                    cand.add(v);
                }
                if(flag){
                    System.out.print(cand.get(0));
                    for(int j = 1; j < cand.size(); j++)
                        System.out.print(" " + cand.get(j));
                    System.out.println();
                    continue loop;
                }
            }
            System.out.println("Impossible");
        }
    }
}

File: pku/2466.cc
#include <stdio.h>
#include <algorithm>
#include <vector>
#include <map>

using namespace std;

int main(int argc, char* argv[]){
  int n;
  for(;~scanf("%d", &n);){
    int total = n * (n - 1) / 2;
    vector<int> ps;
    for(int i = 0; i < total; i++){
      int v;
      scanf("%d", &v);
      ps.push_back(v);
    }
    sort(ps.begin(), ps.end());
    bool solved = false;
    for(int i = 2; i < total; i++){
      vector<int> cand;
      int a2 = ps[0] + ps[1] - ps[i];
      if(a2 % 2 == 1)continue;
      int a = a2 / 2;
      cand.push_back(a);
      cand.push_back(ps[0] - a);
      cand.push_back(ps[1] - a);
      bool flag = true;
      map<int, int> m;
      for(int j = 2; j < total; j++)
        if(j != i){
          if(m.count(ps[j]) > 0)
            m[ps[j]]++;
          else
            m.insert(pair<int, int>(ps[j], 1));
        }
      for(int j = 3; j < n; j++){
        int mn = 1<<30;
        for(int k = 2; k < total; k++)
          if(m.count(ps[k]) > 0 && m[ps[k]] > 0){mn = ps[k]; break;}
        m[mn]--;
        int v = mn - a;
        bool found = true;
        for(int k = 1; k < cand.size(); k++){
          int w = v + cand[k];
          if(m.count(w) == 0 || m[w] <= 0){
            found = false;
            break;
          }
          m[w]--;
        }
        if(!found){
          flag = false;
          break;
        }
        cand.push_back(v);
      }
      if(flag){
        printf("%d", cand[0]);
        for(int i = 1; i < n; i++)
          printf(" %d", cand[i]);
        printf("\n");
        solved = true;
        break;
      }
    }
    if(!solved){
      printf("Impossible\n");
    }
  }

  return 0;
}
